### PR TITLE
Model Account Service as singleton with ability to provide multiple Identity Providers

### PR DESCRIPTION
**Pull request by Supereg**
# Model Account Service as singleton with ability to provide multiple Identity Providers

## :recycle: Current situation & Problems
The current design of SpeziAccount allows to have multiple `AccountService`s that each implement exactly one identity provider (e.g., signup with password credentials or SSO using Sign in with Apple). We are currently missing support for a single `AccountService` to combine multiple identity providers. Secondly, we found that there is generally no need for an application to support multiple, entirely different account services.

Further, the current infrastructure is relatively strict on how to define account services and their UI components. This is done through multiple different sub-protocols to the based `AccountService` protocol. Each protocol makes certain assumptions how an account service might look and is, therefore, (magically) providing certain default implementations or default UI components. While the defaults are great, we found that this design is pretty inflexible. If an account service doesn't quite fit the requirements of one of the default protocols, it has to implement a lot of the things completely from scratch. Most of the default UI components shipped with SpeziAccount are tightly coupled with these protocols and can therefore not be easily reused. We would love a design that provides great defaults, but not at the cost of flexibility.

Further, account notifications and external storage are currently realized using `Standard` constraints. Specifically with external storage we found that these are often provided by external components and that the constraint was always just used to forward the implementation to such a component. Removing the constraint and directly supplying the external storage provider to the `AccountConfiguration` would be much more ideal. In an attempt to transparently support notifications and external storage, internal wrapper account services were used to automatically implement support for these operations. This design required several workarounds when it was required to interact with the underlying account service and was easy to get wrong. Not ideal.

Lastly, there was typically a lot of boiler plate code required to create new account keys, making it a bigger effort than necessary. Notably, creating non-String-based account keys involved creating display and entry view components as there wasn't anything provided out of the box for most primitive types. All in all, it made fast prototyping harder as considerable effort had to be spent when implementing new account keys.

The release notes sections lists how all of these problems have been addressed in this PR.

## :gear: Release Notes 
* `AccountService` is now a singleton design that is capable of providing multiple `IdentityProvider`
  * All previous sub-protocols have been removed and there are only minimal requirements for the `AccountService` protocol.
  * The setup flow for `IdentityProvider` can be completely customized by the account service.
  * There are great default UI components with great customization possibilities for login and signup like the `AccountSetupProviderView` (previously the `UserIdPasswordEmbeddedView`) and a `SignInWithApple` view that integrates nicely with SpeziAccount
*  Account Notifications are now handled through the new `AccountNotifications` module. The `.deletingAccount` event must now be manually emitted by an `AccountService`.
* The new `ExternalAccountStorage` module serves as the bridge between an `AccountService` and an `AccountStorageProvider`.
  * A account service that doesn't support storing arbitrary account details must manually integrate with a storage provider through the `ExternalAccountStorage` module.
  * An `ExternalAccountStorage` can now communicate back changes from the external data store.
  * For both account services and external storages, great example implementations are provided through the `InMemoryAccountService` and the `InMemoryAccountStorageProvider`. Both are great for SwiftUI previews and UI testing.
  * The old, wrapper-based approach was removed.
* There is a new `AccountKey` and `KeyEntry` macro that can be used to define new account keys.
  * Additionally, a lot of display and entry views are now provided by default for most primitive types (e.g., Bool, Numerics, Strings and enum values)
* In general, this PR tried to simplify the components provided by SpeziAccount. This included simplifying the `AccountDetails` storage infrastructure, removing the differentiating between `SignupDetails`, `ModifiedAccountDetails`, `PartialAccountDetails` and `RemovedAccountDetails`. All were replaced by the single `AccountDetails` container. The `AccountValuesBuilder` was also replaced by making all accessed to `AccountDetails` mutable.
* The `FollowUpInfoSheet` is now much more customizable, allowing to change behavior in `AccountSetup` using the `FollowUpBehavior` configuration. See https://github.com/StanfordSpezi/SpeziAccount/pull/53#issuecomment-2273202062.
  * The `verifyRequiredAccountDetails(_:)` modifier was removed as it is now automatically added through the `AccountConfiguration`.

## :books: Documentation
The documentation catalog was restructured. Generally, nesting was reduced and it aims to move types that are relevant to framework users to the root documentation page to make it more easily navigable and discoverable.
Documentation that is tailored towards developers implementing account services or external storage providers got their own group. Generally, all documentation was updated to reflect the new functionality.

## :white_check_mark: Testing
Existing tests were used to verify existing behavior. Some new tests were added to tests new functionalities.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Supereg** [comment]
@PSchmiedmayer markdown lint is currently failing as we are linking to the `FirebaseAccountService` that is not yet deployed in the documentation catalog.

File: Sources/SpeziAccount/AccountValue/Keys/EmailAddressKey.swift [hunk 1]
```diff
@@ -11,65 +11,88 @@ import SpeziValidation
...
-            VerifiableTextField(Key.name, text: $email)
-                .textContentType(.emailAddress)
-                .disableFieldAssistants()
+        guard let configuration = repository[AccountDetails.AccountServiceConfigurationDetailsKey.self],
+              case .emailAddress = configuration.userIdConfiguration.idType else {
+            return nil
         }
+
+        // return the userId if it's a email address
+        return repository[AccountKeys.userId]
```
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/AccountValue/Keys/EmailAddressKey.swift:96]: Nice mechanism!

File: Sources/SpeziAccount/AccountValue/Keys/PersonNameKey.swift [hunk 2]
```diff
@@ -12,114 +12,94 @@ import SpeziViews
...
-        private var nameIsRequired: Bool {
-            account.configuration[Key.self]?.requirement == .required
+            GridValidationStateFooter(familyNameValidation.allDisplayedValidationResults)
         }
-        
-        private var validationRule: ValidationRule {
-            nameIsRequired ? .nonEmpty : .acceptAll
+        .environment(\.validationConfiguration, .considerNoInputAsValid)
+        .onChange(of: givenNameValidation) {
+            print("given Name: \(givenNameValidation)")
```
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/AccountValue/Keys/PersonNameKey.swift:76]: Wondering if we want to keep the print statements and others in here?

File: Sources/SpeziAccount/Environment/FollowUpBehavior.swift [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+//
+
+import SwiftUI
+
+
+/// Determine how the Follow-Up information sheet is presented after account setup.
+///
+/// Use the ``SwiftUI/View/followUpBehaviorAfterSetup(_:)`` modifier to set how the follow-up information sheet is
+/// shown inside the ``AccountSetup`` after a successful setup (login or signup).
+public enum FollowUpBehavior {
```
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/Environment/FollowUpBehavior.swift:16]: Nice configuration option!

File: Sources/SpeziAccount/Environment/SignupProviderCompliance.swift [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        if let next, next.date >= previous.date {
+            value = next
+        }
+    }
+}
+
+
+/// Define the compliance level of a custom signup provider.
+///
+/// SpeziBluetooth imposes requirements on account keys using ``AccountValueConfiguration`` which is supplied by the framework user.
```
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/Environment/SignupProviderCompliance.swift:44]:
```suggestion
/// SpeziAccount imposes requirements on account keys using ``AccountValueConfiguration`` which is supplied by the framework user.
```

File: Sources/SpeziAccount/Environment/SignupProviderCompliance.swift [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+
+extension SignupProviderCompliance: Sendable, Equatable {}
+
+
+extension View {
+    /// Informs the parent account setup view about compliance of the signup provider.
+    ///
+    /// SpeziBluetooth imposes requirements on account keys using ``AccountValueConfiguration`` which is supplied by the framework user.
```
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/Environment/SignupProviderCompliance.swift:119]:
```suggestion
    /// SpeziAccount imposes requirements on account keys using ``AccountValueConfiguration`` which is supplied by the framework user.
```

File: Sources/SpeziAccount/Mock/InMemoryAccountService.swift
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/Mock/InMemoryAccountService.swift:1]: Nice example and reference implementation; great to see how these things are coming together here.

File: Sources/SpeziAccount/Mock/InMemoryAccountStorageProvider.swift
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/Mock/InMemoryAccountStorageProvider.swift:1]: Nice compact implementation!

File: Sources/SpeziAccount/SpeziAccount.docc/AccountKey/Adding new Account Values.md [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+it might be necessary or improve the user experience to provide your own user interfaces.
+
+This is done by implementing a ``DataDisplayView`` or ``DataEntryView`` respectively.
+
+- Note: The <doc:User-Interfaces-that-are-provided-by-default> section provides an overview when SpeziAccount is able to provide default user interfaces.
+
+The code example below implements a custom `EntryView` and `DisplayView` and updates the `AccountKey` declaration from above to use
+the new views.
+
+```swift
```
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/SpeziAccount.docc/AccountKey/Adding new Account Values.md:64]: Cool example, nice job!

File: Sources/SpeziAccount/SpeziAccount.docc/AccountKey/Adding new Account Values.md
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/SpeziAccount.docc/AccountKey/Adding new Account Values.md:1]: Nice summary here!

File: Sources/SpeziAccount/SpeziAccount.docc/AccountService/Creating your own Account Service.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+private struct MySetupView: View {
+    @Environment(MyAccountService.self)
+    private var service
+
+    var body: some View {
+        // An AccountSetupProviderView that doesn't support password reset functionality.
+        AccountSetupProviderView { credentials in
+            try await service.login(userId: credentials.userId, password: credentials.password)
+        } signup: { details in
+            // try await service.signup(details)
```
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/SpeziAccount.docc/AccountService/Creating your own Account Service.md:137]: Is that commented out by design?

File: Sources/SpeziAccount/SpeziAccount.docc/Resources/AccountOverview.png
Comment (PSchmiedmayer) [inline Sources/SpeziAccount/SpeziAccount.docc/Resources/AccountOverview.png:1]: Nice idea with the icons in the section headings. As far as I can see we want to indicate that these are sections with detail views that require navigation while elements without icons do not and don't have an icon?

File: Sources/SpeziAccountMacros/AccountKeyMacro.swift
Comment (PSchmiedmayer) [inline Sources/SpeziAccountMacros/AccountKeyMacro.swift:1]: Amazing to see the infrastructure needed for a macro; definitely not a nice progressive disclosure of language features.

File: Tests/SpeziAccountMacrosTests/AccountKeyMacroTests.swift
Comment (PSchmiedmayer) [inline Tests/SpeziAccountMacrosTests/AccountKeyMacroTests.swift:1]: Amazing; thanks for all the tests here!

File: Tests/SpeziAccountTests/__Snapshots__/SnapshotTesting/testAccountProviderViewLayoutVariations.iphone-view0.png
Comment (PSchmiedmayer) [inline Tests/SpeziAccountTests/__Snapshots__/SnapshotTesting/testAccountProviderViewLayoutVariations.iphone-view0.png:1]: I am wondering if we can take some UI inspiration from out QDG username & password fields and the option to show/hide the password using a button as an functionality that we could move into Spezi Account in the long run & future PR.

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for the amazing work here @Supereg!!

I had a few small comments throughout the PR; most of them should be quick to address and some of them might be addressed in the future in follow-up PRs as we bake the 2.0 version in beta releases.

Thanks for all the work; really great to see how this is converging into the current setup!

File: .swiftlint.yml [hunk 8]
```diff
@@ -367,9 +367,6 @@ only_rules:
   # The variable should be placed on the left, the constant on the right of a comparison operator.
   - yoda_condition
 
-attributes:
-  attributes_with_arguments_always_on_line_above: false
-
 deployment_target: # Availability checks or attributes shouldn’t be using older versions that are satisfied by the deployment target.
   iOSApplicationExtension_deployment_target: 16.0
```
[the remark this answers is not in the document]
Comment (Supereg) [reply .swiftlint.yml:371]: Remove the rule overrides 👍

File: Sources/SpeziAccount/AccountValue/Keys/PersonNameKey.swift (hunk 2 above)
Comment (Supereg) [reply Sources/SpeziAccount/AccountValue/Keys/PersonNameKey.swift:76]: Great catch 👍

File: Sources/SpeziAccount/Environment/SignupProviderCompliance.swift (hunk 4 above)
Comment (Supereg) [reply Sources/SpeziAccount/Environment/SignupProviderCompliance.swift:44]: Upsie, thanks 👍

File: Sources/SpeziAccount/SpeziAccount.docc/AccountService/Creating your own Account Service.md (hunk 7 above)
Comment (Supereg) [reply Sources/SpeziAccount/SpeziAccount.docc/AccountService/Creating your own Account Service.md:137]: Not sure how the line got commented out, removed the double slashes, thanks 👍

File: Sources/SpeziAccount/SpeziAccount.docc/Resources/AccountOverview.png
Comment (Supereg) [reply Sources/SpeziAccount/SpeziAccount.docc/Resources/AccountOverview.png:1]: The icons don't really have a semantic apart from representing the element. Just tried to make the view look nicer.

File: Tests/SpeziAccountTests/__Snapshots__/SnapshotTesting/testAccountProviderViewLayoutVariations.iphone-view0.png
Comment (Supereg) [reply Tests/SpeziAccountTests/__Snapshots__/SnapshotTesting/testAccountProviderViewLayoutVariations.iphone-view0.png:1]: Yes, I still have the show/hide password field from QDG on my list. Will create an issue though and include it in a future PR.

Comment (PSchmiedmayer) [reply Tests/SpeziAccountTests/__Snapshots__/SnapshotTesting/testAccountProviderViewLayoutVariations.iphone-view0.png:1]: Thank you!

File: Sources/SpeziAccount/SpeziAccount.docc/Resources/AccountOverview.png
Comment (PSchmiedmayer) [reply Sources/SpeziAccount/SpeziAccount.docc/Resources/AccountOverview.png:1]: Makes sense; it looks really cool. 

I am just wondering about consistency with all other edit fields and sections . Should we use that for other sections and/or make them navigation links? Not something to change here but to think about in the next meeting.

File: Package.swift [hunk 9]
```diff
@@ -23,20 +24,36 @@ let package = Package(
     name: "SpeziAccount",
     defaultLocalization: "en",
     platforms: [
-        .iOS(.v17)
+        .iOS(.v17),
+        .macOS(.v14) // we need to specify that to run macro tests as they only run on the host platform
```
[the remark this answers is not in the document]
Comment (Supereg) [reply Package.swift]: Added visionOS as supported platform and make sure tests run on visionOS.

File: Sources/SpeziAccount/AccountOverview.swift [hunk 10]
```diff
@@ -42,26 +39,49 @@ import SwiftUI
...
+    /// Defines the behavior for the close button.
+    public enum CloseBehavior {
+        /// No close button is shown.
+        case disabled
+        /// A close button is shown that calls the `dismiss` action.
+        case showCloseButton
+    }
+
+    /// Defines the behavior of delete functionality.
+    public enum AccountDeletionBehavior {
```
[the remark this answers is not in the document]
Comment (Supereg) [reply Sources/SpeziAccount/AccountOverview.swift:71]: Added a new `belowLogout` option that shows both buttons, always.

[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply Sources/SpeziAccount/AccountOverview.swift:71]: Thank you!

**Comment by PSchmiedmayer** [comment]
Great job to get this merged @Supereg; a huge lift!

**Pull request merged**